deps/eventlog/src/ccel/rtmr.rs

```diff
@@ -0,0 +1,168 @@
+use crate::CcEventLog;
+use crate::TcgAlgorithm;
+use anyhow::{bail, Result};
+use sha2::{Digest, Sha256, Sha384, Sha512};
+use std::collections::HashMap;
+use std::result::Result::Ok;
+
+const RTMR_LENGTH_BY_BYTES: usize = 48;
+const CHECK_RTMR_LIMIT: usize = 3;
+
+#[derive(Debug, Clone)]
+pub struct Rtmr {
+    pub data: [Vec<u8>; 4],
+}
+
+impl Rtmr {
+    pub fn integrity_check(&self, rtmr_from_quote: [Vec<u8>; 4]) -> Result<()> {
+        for (index, quote_value) in rtmr_from_quote.iter().enumerate().take(CHECK_RTMR_LIMIT) {
+            let ccel_value = &self.data[index];
+            if ccel_value != quote_value {
+                bail!(
+                    "CCEL eventlog does not pass RTMR [{}] check. CCEL value: {}, Quote value: {}",
+                    index,
+                    hex::encode(ccel_value),
+                    hex::encode(quote_value)
+                );
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl TryFrom<CcEventLog> for Rtmr {
+    type Error = anyhow::Error;
+
+    fn try_from(data: CcEventLog) -> anyhow::Result<Self> {
+        let mut result: HashMap<u32, [u8; RTMR_LENGTH_BY_BYTES]> = HashMap::new();
+
+        for entry in data.log.into_iter() {
+            let digest = &entry.digests[0].digest;
+
+            let mr_value = result
+                .entry(entry.index)
+                .or_insert([0u8; RTMR_LENGTH_BY_BYTES]);
+
+            let hash = accumulate_hash(
+                entry.digests[0].alg,
+                mr_value.clone().to_vec(),
+                digest.as_slice(),
+            )?;
+
+            mr_value.copy_from_slice(&hash);
+        }
+
+        let mut data: [Vec<u8>; 4] = [
```

Comment:
we cannot assume it's 4 anymore. See the latest UEFI spec for other vendor details. with that, would make sense to move the replay logic to the caller side which the measurement registry count and the registry width